include/blast/math/register_matrix/RegisterMatrix.hpp

```diff
@@ -738,54 +767,116 @@ namespace blast
+                        v_[ii][j] = fmadd(ax[ii], bx, v_[ii][j]);
+                }
+
+                a.hmove(1);
+                bu.vmove(1);
             }
+        }
+        else
+        {
+            BLAST_THROW_EXCEPTION(std::logic_error {"Left multiplication with lower-triangular matrix not implemented in RegisterMatrix::trmm()"});
+        }
+    }
 
-            a.hmove(1);
-            bu.vmove(1);
+
+    template <typename T, size_t M, size_t N, bool SO>
+    template <typename PB, typename PA>
+    requires MatrixPointer<PB, T> && (PB::storageOrder == columnMajor) && MatrixPointer<PA, T>
+    BLAZE_ALWAYS_INLINE void RegisterMatrix<T, M, N, SO>::trmm(T alpha, PB b, PA a, UpLo uplo, bool diagonal_unit) noexcept
+    {
+        if (diagonal_unit)
+            BLAST_THROW_EXCEPTION(std::logic_error {"Unit diagonal matrices support not implemented in RegisterMatrix::trmm()"});
+
+        if (uplo == UpLo::Lower)
+        {
+            auto au = ~a;
+
+            if constexpr (SO == columnMajor)
+            {
+                #pragma unroll
+                for (size_t k = 0; k < N; ++k)
+                {
+                    SimdVecType bx[RM];
+
+                    #pragma unroll
+                    for (size_t i = 0; i < RM; ++i)
+                        bx[i] = alpha * b(i * SS, 0).load();
+
+                    #pragma unroll
+                    for (size_t j = 0; j <= k; ++j)
+                    {
+                        SimdVecType ax = au(0, j).broadcast();
+
+                        #pragma unroll
+                        for (size_t i = 0; i < RM; ++i)
+                            v_[i][j] = fmadd(bx[i], ax, v_[i][j]);
+                    }
+
+                    b.hmove(1);
+                    au.vmove(1);
+                }
+            }
+            else
+            {
+                BLAZE_THROW_LOGIC_ERROR("Not implemented");
+            }
+        }
+        else
+        {
+            BLAST_THROW_EXCEPTION(std::logic_error {"Right multiplication with upper-triangular matrix not implemented in RegisterMatrix::trmm()"});
         }
     }
 
 
     template <typename T, size_t M, size_t N, bool SO>
     template <typename PB, typename PA>
     requires MatrixPointer<PB, T> && (PB::storageOrder == columnMajor) && MatrixPointer<PA, T>
-    BLAZE_ALWAYS_INLINE void RegisterMatrix<T, M, N, SO>::trmmRightLower(T alpha, PB b, PA a) noexcept
+    BLAZE_ALWAYS_INLINE void RegisterMatrix<T, M, N, SO>::trmm(T alpha, PB b, PA a, UpLo uplo, bool diagonal_unit, size_t m, size_t n) noexcept
     {
+        // NOTE: this implementation does uses unmasked loads from the matrix a,
+        // and therefore will access rows of a beyond m-1.
+        // This will result in undefined behavior on unpadded matrices.
         auto au = ~a;
 
         if constexpr (SO == columnMajor)
         {
             #pragma unroll
-            for (size_t k = 0; k < N; ++k)
+            for (size_t k = 0; k < N; ++k) if (k < n)
```

Comment:
Formatting

Reply:
This is intentional. If is my way to write an unrolled loop from 0 to `n`.